### PR TITLE
feat(bulk-load): add perf-counter

### PR DESCRIPTION
**Pull request by hycdong**
This pull request add bulk load perf-counters, including:
```
perf_counter_wrapper _counter_bulk_load_running_count;
perf_counter_wrapper _counter_bulk_load_downloading_count;
perf_counter_wrapper _counter_bulk_load_ingestion_count;
perf_counter_wrapper _counter_bulk_load_succeed_count;
perf_counter_wrapper _counter_bulk_load_failed_count;
perf_counter_wrapper _counter_bulk_load_ingestion_reject_write_count;
perf_counter_wrapper _counter_bulk_load_download_file_succ_count;
perf_counter_wrapper _counter_bulk_load_download_file_fail_count;
perf_counter_wrapper _counter_bulk_load_download_file_size;
perf_counter_wrapper _counter_bulk_load_max_ingestion_time_ms;
perf_counter_wrapper _counter_bulk_load_max_duration_time_ms;
```
#### bulk_load_running_count
name: `replica*eon.replica_stub*bulk.load.running.count`
replica count who are executing bulk load

#### bulk_load_downloading_count
name: `replica*eon.replica_stub*bulk.load.downloading.count`
replica count who are executing bulk load downloading

#### bulk_load_ingestion_count
name: `replica*eon.replica_stub*bulk.load.ingestion.count`
replica count who are executing bulk load ingesting

#### bulk_load_succeed_count
name: `replica*eon.replica_stub*bulk.load.succeed.count`
replica count who bulk load succeed

#### bulk_load_failed_count
name: `replica*eon.replica_stub*bulk.load.failed.count`
replica count who bulk load failed

#### bulk_load_ingestion_reject_write_count
name: `replica*eon.replica_stub*bulk.load.ingestion.reject.write.count`
the count of rejected write requests during bulk load ingestion

#### bulk_load_download_file_succ_count
name: `replica*eon.replica_stub*bulk.load.download.file.success.count`
download file succeed count during bulk load downloading

#### bulk_load_download_file_fail_count
name: `replica*eon.replica_stub*bulk.load.download.file.fail.count`
download file failed count during bulk load downloading

#### bulk_load_download_file_size
name: `replica*eon.replica_stub*bulk.load.download.file.size`
downloaded file size during bulk load downloading

#### bulk_load_max_ingestion_time_ms
name: `replica*eon.replica_stub*bulk.load.max.ingestion.duration.time.ms`
bulk load max ingestion duration time(ms)

#### bulk_load_max_duration_time_ms
name: `replica*eon.replica_stub*bulk.load.max.duration.time.ms`
bulk load max duration time(ms)

### CODE REVIEW

File: src/replica/replica_stub.h [hunk 1]
```diff
@@ -435,6 +435,20 @@ class replica_stub : public serverlet<replica_stub>, public ref_counter
...
     perf_counter_wrapper _counter_tcmalloc_release_memory_size;
 #endif
+
+    perf_counter_wrapper _counter_bulk_load_running_count;
+    perf_counter_wrapper _counter_bulk_load_downloading_count;
+    perf_counter_wrapper _counter_bulk_load_ingestion_count;
+    perf_counter_wrapper _counter_bulk_load_finish_count;
+    perf_counter_wrapper _counter_bulk_load_failed_count;
+    perf_counter_wrapper _counter_bulk_load_ingestion_reject_write_count;
+    perf_counter_wrapper _counter_bulk_load_recent_download_file_succ_count;
```
Comment (acelyc111) [inline src/replica/replica_stub.h:445]: What does 'recent' mean? I didn't found the differene to the one not named with 'recent'.

File: src/replica/replica_stub.cpp [hunk 2]
```diff
@@ -1716,6 +1772,10 @@ void replica_stub::on_gc()
     uint64_t cold_backup_running_count = 0;
     uint64_t cold_backup_max_duration_time_ms = 0;
     uint64_t cold_backup_max_upload_file_size = 0;
+    uint64_t bulk_load_running_count = 0;
```
Comment (acelyc111) [inline src/replica/replica_stub.cpp:1775]: It's not this PR's problem, but these statistics are under function 'on_gc' seems strange. Could you please refactor it by another PR?

File: src/replica/replica_stub.cpp [hunk 3]
```diff
@@ -335,6 +335,62 @@ void replica_stub::install_perf_counters()
         COUNTER_TYPE_VOLATILE_NUMBER,
         "write size exceed threshold count in the recent period");
 
+    // <- Bulk Load Metrics ->
+
+    _counter_bulk_load_running_count.init_app_counter("eon.replica_stub",
+                                                      "bulk.load.running.count",
+                                                      COUNTER_TYPE_VOLATILE_NUMBER,
+                                                      "current bulk load running count");
+    _counter_bulk_load_downloading_count.init_app_counter("eon.replica_stub",
```
Comment (acelyc111) [inline src/replica/replica_stub.cpp:344]: This metric is a GAUGE type, right? If it is, `decrease` is missing?

Comment (hycdong) [reply src/replica/replica_stub.cpp:344]: https://github.com/XiaoMi/rdsn/blob/95002ff5306b835910a986ddfb636811488e3e3b/include/dsn/perf_counter/perf_counter.h#L36-L43
`COUNTER_TYPE_VOLATILE_NUMBER` is a special type, it needn't decrease the value.

File: src/replica/replica_stub.h (hunk 1 above)
Comment (hycdong) [reply src/replica/replica_stub.h:445]: As the comment below, the value of those perf_counters will be reset while getting the value, actually all of them it the `recent` value during this reporting round, not the total value.

File: src/replica/replica_stub.cpp (hunk 2 above)
Comment (hycdong) [reply src/replica/replica_stub.cpp:1775]: Okay, I will try to refactor it further.

File: src/replica/replica_stub.h (hunk 1 above)
Comment (acelyc111) [reply src/replica/replica_stub.h:445]: So how about to remove `recent` to reduce readers' confusion？

Comment (hycdong) [reply src/replica/replica_stub.h]: done

**Comment by neverchanje** [comment]
You need to give a list of the full names of these perf-counters, not only the variable names. Refer to this PR https://github.com/XiaoMi/rdsn/pull/393 if you're not familiar with the information you need to provide.

**Comment by hycdong** [comment]
> You need to give a list of the full names of these perf-counters, not only the variable names. Refer to this PR #393 if you're not familiar with the information you need to provide.

I add some simple descriptions, if you think any perf-counter making you confused, I can add more details about it.

File: src/replica/replica_stub.cpp [hunk 4]
```diff
@@ -335,6 +335,61 @@ void replica_stub::install_perf_counters()
...
+    _counter_bulk_load_download_file_fail_count.init_app_counter(
+        "eon.replica_stub",
+        "bulk.load.download.file.fail.count",
+        COUNTER_TYPE_VOLATILE_NUMBER,
+        "bulk load recent download file failed count");
+    _counter_bulk_load_download_file_size.init_app_counter("eon.replica_stub",
+                                                           "bulk.load.download.file.size",
+                                                           COUNTER_TYPE_VOLATILE_NUMBER,
+                                                           "bulk load recent download file size");
+    _counter_bulk_load_max_download_file_size.init_app_counter("eon.replica_stub",
```
Comment (neverchanje) [inline src/replica/replica_stub.cpp:379]: What is a "max" file size used for? Why not calculate this value by `COUNTER_TYPE_PERCENTILE`?

Comment (hycdong) [reply src/replica/replica_stub.cpp:379]: It is used to record max downloaded file size, just to show the maximum the size. It seems that it is a little meaningless.

**Pull request merged**